cpp/src/ray/runtime/abstract_ray_runtime.cc

```
@@ -0,0 +1,118 @@
+
+#include "abstract_ray_runtime.h"
+
+#include <cassert>
+
+#include <ray/api.h>
+#include <ray/api/ray_mode.h>
+#include "../address_helper.h"
+#include "ray_dev_runtime.h"
+#include "ray_native_runtime.h"
+#include "task/invocation_executor.h"
+
+namespace ray {
+namespace api {
+
+std::unique_ptr<AbstractRayRuntime> AbstractRayRuntime::ins_;
+std::once_flag AbstractRayRuntime::isInited_;
+
+AbstractRayRuntime &AbstractRayRuntime::Init(std::shared_ptr<RayConfig> config) {
+  DoInit(config);
+  Ray::Init();
+  return *ins_;
+}
+
+AbstractRayRuntime &AbstractRayRuntime::DoInit(std::shared_ptr<RayConfig> config) {
+  std::call_once(isInited_, [config] {
+    if (config->runMode == RunMode::SINGLE_PROCESS) {
+      ins_.reset(new RayDevRuntime(config));
+      AddressHelperInit();
+    } else {
+      ins_.reset(new RayNativeRuntime(config));
```

I think we can just throw an exception here.